docs/proposals/blocks-storage-sharding.md

```diff
@@ -0,0 +1,84 @@
+
+### Query execution
+
+When a querier executes a query, it will need to fetch series both from ingesters and the store-gateway instances.
+
+For a given query, the number of blocks to query is expected to be low, especially if the Cortex cluster is running the `query-frontend` with a `24h` query split interval. In this scenario, whatever is the client's query time range, the `query-frontend` will split the client's query into sub-queries each with up to `24h` time range and the querier will likely hit not more than 1 block per sub-query (except for the last 24h for which blocks may have not been compacted yet).
+
+Given this assumption, we want to avoid sending every query to every store-gateway instance. The querier should be able to take an informed decision about the minimum subset of store-gateway instances which needs to query given a time range.
+
+The idea is to run the [`MetaFetcher`](https://github.com/thanos-io/thanos/blob/master/pkg/block/fetcher.go#L127) also within the querier, but without any sharding filter (contrary to the store-gateway). At any given point in time, the querier knows the entire list of blocks in the storage. When the querier executes the `Select()` (or `SelectSorted()`) it does:
+
+1. Compute the list of blocks by the query time range
+2. Compute the minimum list of store-gateway instances containing the required blocks (using the information from the ring)
+3. Fetch series from ingesters and the matching store-gateway instances
+4. Merge and deduplicate received series
+   - Optimization: can be skipped if the querier hits only 1 store-gateway
+
+### Query results consistency
+
+When a querier executes a query, it should guarantee that either all blocks matching the time range are queried or the query fails.
+
+However, due to the (intentional) lack of a strong coordination between queriers and store-gateways, and the ring topology which can change any time, there's no guarantee that the blocks assigned to a store-gateway shard are effectively loaded on the store-gateway itself at any given point in time.
+
+The idea is introduce a **consistency check in the querier**. When a store-gateway receives a request from the querier, the store-gateway includes in the response the list of block IDs currently loaded on the store-gateway itself. The querier can then merge the list of block IDs received from all store-gateway hit, and match it against the list of block IDs computed at the beginning of the query execution.
+
+There are three possible scenarios:
+
+1. The list match: all good
+2. All the blocks known by the querier are within the list of blocks returned by store-gateway, but the store-gateway also included blocks unknown to the querier: all good (it means the store-gateways have discovered and loaded new blocks before the querier discovered them)
+3. Some blocks known by the querier are **not** within the list of blocks returned by store-gateway: potential consistency issue
```

Comment:
There is also 4th case: Since you look on response result only you need to first request things from it to tell... That's why `Info` method  might be a better fit

Reply:
> That's why Info method might be a better fit

We considered using the `Info()` which is basically an API exposed by the store-gateway. However, even if you query all store-gateways to fetch all their `Info()` there's no guarantee that all blocks are loaded across by the running gateways at any given point in time (because of how the re-sharding works whenever the ring topology changes, which hasn't any sort of "distributed lock" around it). For this reason, the proposal is about having the querier keeping an independent list of blocks (without any filter) and to run a consistency check on the querier side.

Reply:
But that is not the design. You also propose something like Info from result, which is weird.

Reply:
In order to run the consistency check on the querier side, we need to know (on the querier side) which block have been hit for a given `Series()` API call in the `store-gateway`. This is the reason why the `Series()` response is expected to also list the block IDs hit by the request, so that the querier can merge all block IDs contained in `store-gateway` responses (in case 2+ gateway have been hit) and match against the list of blocks computed on the querier side (based on the known blocks by the `MetaFetcher` running within the querier itself).